tests/rptest/tests/describe_topics_test.py

```diff
@@ -167,38 +148,87 @@ def test_describe_topics_with_documentation_and_types(self):
             )
         }
 
-        for line in property_lines:
-            col = line.strip().lower().split()
-            self.logger.debug(col)
-            assert len(col) >= 4
-            name = col[0]
-            config_type = col[1]
-            value = col[2]
-            source_type = col[3]
-
-            # The above .split() also applies to the value for datapolicy
-            # so we need to re-concatenate the string
-            if name == "redpanda.datapolicy":
-                assert len(col) == 5
-                value = f'{col[2]} {col[3]}'
-                source_type = col[4]
-
-            assert name in properties.keys()
+        tp_spec = TopicSpec()
+        self.client().create_topic([tp_spec])
+        kcl = KCL(self.redpanda)
+        output = kcl.describe_topic(tp_spec.name,
+                                    with_docs=True,
+                                    with_types=True).lower().splitlines()
+
+        # The output format from topic describe has the following structure
+        #
+        # - A table section with descriptions for the property name, type, value, and source
+        # - An empty line
+        # - Three lines for each property, called the documentation section:
+        #     1. Property name with a colon (e.g., cleanup.policy:)
+        #     2. Property documentation string
+        #     3. An empty line
+
+        # First check redpanda.datapolicy in the property table because the value is a special case.
+        # The extra space is necessary so the datapolicy lines in the documentation section are not removed.
+        copro_lines = [
+            line for line in output if line.startswith('redpanda.datapolicy ')
+        ]
+        self.logger.debug(f"Copro lines {copro_lines}")
+        assert len(copro_lines) == 1
+        copro_re = re.compile(
+            r"^(?P<name>redpanda.datapolicy)\s+(?P<type>string)\s+(?P<value>function_name:  script_name:)\s+(?P<src>default_config)$"
+        )
+        copro_match = copro_re.match(copro_lines[0])
+        assert copro_match is not None
+
+        all_other_lines = [
+            line for line in output
+            if not line.startswith('redpanda.datapolicy ')
+        ]
+        all_other_iter = iter(all_other_lines)
+        property_re = re.compile(
+            r"^(?P<name>[a-z.]+?)\s+(?P<type>[a-z]+?)\s+(?P<value>\S+?)\s+(?P<src>[a-z_]+?)$"
+        )
+
+        line = next(all_other_iter)
+        self.logger.debug(f"Property line {line}")
+        prop_match = property_re.match(line)
+        while prop_match is not None:
```

Comment:
Maybe it would be better to loop over all lines (so you have a definitive stop condition, i.e. when there are no more lines) and use a `break` statement to exit early when you observe `prop_match` is `None`.

Thinking about possible infinite loops and or handling exceptions when `next()` reaches the end of the input might be harder to reason about then just a simple `for line in lines` loop